File: iosApp/iosApp/Components/BalanceDisplay.swift
import Shared
import SwiftUI

/// The PWA's `BalanceDisplay` (U18, KTD-11, R12): unified total as a BIP177
/// `₿` amount in the display font, a `+₿X pending` line, and a hide/show
/// toggle. Visibility is persisted by the caller under the PWA's
/// `balance-visible` key; hidden renders six dots. The readout scales down
/// past 5 digits (the PWA's clamp equivalent: text-7xl → text-5xl).
///
/// `FormatKt` is one of the two UI-safe pure helpers views may consume from
/// the shared framework directly (R14 carve-out) — identical formatting on
/// both clients comes from the same commonMain code.
struct BalanceDisplay: View {
    let balanceSats: Int64
    let visible: Bool
    let onToggleVisible: () -> Void
    var pendingSats: Int64?
    var breakdown: String?
    var loading: Bool = false

    @Environment(\.zinqqColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if loading {
                ProgressView()
                    .tint(colors.onField)
                    .frame(width: 32, height: 32)
                    .accessibilityLabel("Loading balance")
            } else {
                readout
                if let pendingSats, pendingSats > 0, visible {
                    Text("+\(FormatKt.formatBtc(sats: pendingSats)) pending")
                        .font(ZinqqFont.sans(14))
                        .foregroundColor(colors.onFieldMuted)
                        .padding(.top, 4)
                }
                if let breakdown, visible {
                    Text(breakdown)
                        .font(ZinqqFont.sans(14))
                        .foregroundColor(colors.onFieldMuted)
                        .padding(.top, 4)
                }
                toggle
            }
        }
    }

    @ViewBuilder
    private var readout: some View {
        if visible {
            let formatted = FormatKt.formatBtc(sats: balanceSats)
            // Digit-count breakpoint replaces the PWA's vw clamp: 5 digits or
            // fewer read at 72pt, longer amounts drop to 48pt.
            let digits = formatted.filter(\.isNumber).count
            Text(formatted)
                .font(ZinqqFont.display(digits > 5 ? 48 : 72, weight: .bold))
                .kerning(-1)
                .foregroundColor(colors.onField)
                .lineLimit(1)
                .accessibilityLabel("Balance \(formatted)")
        } else {
            Text("••••••")
                .font(ZinqqFont.display(36, weight: .bold))
                .kerning(4)
                .foregroundColor(colors.onField)
                .accessibilityLabel("Balance hidden")
        }
    }

    private var toggle: some View {
        Button(action: onToggleVisible) {
            HStack(spacing: 8) {
                Image(systemName: visible ? "eye.slash" : "eye")
                    .font(.system(size: 16))
                Text(visible ? "Hide balance" : "Show balance")
                    .font(ZinqqFont.sans(14, weight: .medium))
            }
            .foregroundColor(colors.onFieldMuted)
            .frame(minHeight: ZinqqDimens.minTouchTarget)
        }
        .padding(.top, 4)
        .accessibilityLabel(visible ? "Hide balance" : "Show balance")
    }
}

File: androidApp/src/main/kotlin/zinqq/app/components/BalanceDisplay.kt
package zinqq.app.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import zinqq.app.R
import zinqq.app.theme.ZinqqDimens
import zinqq.app.theme.ZinqqTheme
import zinqq.main.formatBtc

/**
 * The PWA's `BalanceDisplay` (U13, KTD-11, R12): unified total as a BIP177
 * `₿` amount in the display font, a `+₿X pending` line, and a hide/show
 * toggle. Visibility is persisted by the caller under the PWA's
 * `balance-visible` key (DataStore); hidden renders six dots. The readout
 * scales down past 5 digits (the PWA's clamp equivalent: text-7xl → text-5xl).
 */
@Composable
fun BalanceDisplay(
    balanceSats: Long,
    visible: Boolean,
    onToggleVisible: () -> Unit,
    modifier: Modifier = Modifier,
    pendingSats: Long? = null,
    breakdown: String? = null,
    loading: Boolean = false,
) {
    val colors = ZinqqTheme.colors
    Column(modifier = modifier, horizontalAlignment = Alignment.Start) {
        if (loading) {
            CircularProgressIndicator(
                color = colors.onField,
                trackColor = colors.onField.copy(alpha = 0.3f),
                strokeWidth = 3.dp,
                modifier = Modifier.size(32.dp),
            )
        } else {
            if (visible) {
                val formatted = formatBtc(balanceSats)
                // Digit-count breakpoint replaces the PWA's vw clamp: 5 digits
                // or fewer read at 72sp, longer amounts drop to 48sp.
                val digits = formatted.count { it.isDigit() }
                Text(
                    text = formatted,
                    color = colors.onField,
                    fontFamily = ZinqqTheme.fonts.display,
                    fontWeight = FontWeight.Bold,
                    fontSize = if (digits > 5) 48.sp else 72.sp,
                    lineHeight = if (digits > 5) 48.sp else 72.sp,
                    letterSpacing = (-1).sp,
                    modifier = Modifier.semantics {
                        contentDescription = "Balance $formatted"
                    },
                )
            } else {
                Text(
                    text = "••••••",
                    color = colors.onField,
                    fontFamily = ZinqqTheme.fonts.display,
                    fontWeight = FontWeight.Bold,
                    fontSize = 36.sp,
                    letterSpacing = 4.sp,
                    modifier = Modifier.semantics { contentDescription = "Balance hidden" },
                )
            }
            if (pendingSats != null && pendingSats > 0 && visible) {
                Text(
                    text = "+${formatBtc(pendingSats)} pending",
                    color = colors.onFieldMuted,
                    fontSize = 14.sp,
                    modifier = Modifier.padding(top = 4.dp),
                )
            }
            if (breakdown != null && visible) {
                Text(
                    text = breakdown,
                    color = colors.onFieldMuted,
                    fontSize = 14.sp,
                    modifier = Modifier.padding(top = 4.dp),
                )
            }
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier
                    .padding(top = 12.dp)
                    .heightIn(min = ZinqqDimens.MinTouchTarget)
                    .clickable(onClick = onToggleVisible)
                    .semantics {
                        contentDescription = if (visible) "Hide balance" else "Show balance"
                    },
            ) {
                Icon(
                    painter = painterResource(
                        if (visible) R.drawable.ic_eye_off else R.drawable.ic_eye,
                    ),
                    contentDescription = null,
                    tint = colors.onFieldMuted,
                    modifier = Modifier.size(20.dp),
                )
                Text(
                    text = if (visible) "Hide balance" else "Show balance",
                    color = colors.onFieldMuted,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                )
            }
        }
    }
}
